### PR TITLE
nixos/vim: add enable option and link `/share/vim-plugins` only when …

### PR DESCRIPTION
**Pull request by jopejoe1**
…vim is enabled

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/programs/vim.nix [hunk 1]
```diff
@@ -4,22 +4,20 @@ let
...
 
     package = lib.mkPackageOption pkgs "vim" {
       example = "vim-full";
     };
   };
 
-  config = lib.mkIf cfg.defaultEditor {
-    environment.systemPackages = [ cfg.package ];
-    environment.variables = { EDITOR = lib.mkOverride 900 "vim"; };
+  config = lib.mkIf cfg.enable {
```
Comment (Aleksanaa) [inline nixos/modules/programs/vim.nix:16]:
```suggestion
  # TODO: convert it into assert
  config = lib.mkIf (cfg.enable || cfg.defaultEditor) (lib.warnIf (cfg.defaultEditor && !cfg.enable) "programs.vim.defaultEditor will only work if programs.vim.enable is enabled" {
```

Not very pleasant, but we should at least make sure the user isn't thrown into a "where is my editor???" situation. I'm not sure how will formatting go here

Comment (jopejoe1) [reply nixos/modules/programs/vim.nix]: Yeah makes sense did not think about that, added a slightly modified version of this.

Comment (jopejoe1) [reply nixos/modules/programs/vim.nix]: Switched over to using `config.warnings` as `lib.warnIf`gave an infinity recursive error.

Comment (Aleksanaa) [reply nixos/modules/programs/vim.nix]: > Switched over to using `config.warnings` as `lib.warnIf`gave an infinity recursive error.

You're right, I forgot how to do this.

**Review by Aleksanaa — APPROVED** [review]

File: nixos/modules/programs/vim.nix [hunk 2]
```diff
@@ -1,25 +1,31 @@
...
+    package = lib.mkPackageOption pkgs "vim" { example = "vim-full"; };
   };
 
-  config = lib.mkIf cfg.defaultEditor {
-    environment.systemPackages = [ cfg.package ];
-    environment.variables = { EDITOR = lib.mkOverride 900 "vim"; };
+  # TODO: convert it into assert after 24.11 release
+  config = lib.mkIf (cfg.enable || cfg.defaultEditor) {
+    warnings = lib.mkIf (cfg.defaultEditor && !cfg.enable) [
+      "programs.vim.defaultEditor will only work if programs.vim.enable is enabled, will be encfored after the 24.11 release"
```
Comment (matdibu) [inline nixos/modules/programs/vim.nix:23]: typo: "encfored"

**Pull request merged**